src/transformStateWithClones.js

```diff
@@ -6,8 +6,35 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let currentState = Object.assign({}, state);
+
+  for (const action of actions) {
+    let newState = Object.assign({}, currentState);
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const keyToRemove of action.keysToRemove) {
+          delete newState[keyToRemove];
+        }
+        break;
+    }
+
+    result.push(newState);
+    currentState = newState;
```

Comment:
[DONT REPEAT YOURSELF] - The action of pushing the newState to the result array is done within the switch statement, which is correct. However, assigning 'newState' to 'currentState' is repeated after every case. This could be moved outside of the switch statement to follow the DRY principle.